### DISCUSSION TITLE
How to change temperature?

### DISCUSSION THREAD

**Discussion by fernandomiletto** [tryAGI/LangChain#366] [Q&A]
I have always used predefined models from OpenAI, particularly Gpt-3.5-Turbo. However, now I am facing a problem that requires a more deterministic model. How can I define a model specifying the desired model, the temperature, and my API key?

**Comment by HavenDV** [comment]
You can use:
- provider level settings: `OpenAiProvider.ChatSettings` property
- model level settings: `OpenAiChatModel.Settings` property
- request level settings: specify settings parameter in `GenerateAsync`

just assign a `new OpenAiChatSettings()` and specify properties to override. It will use more high level setting value or default if not specified.

![image](https://github.com/tryAGI/LangChain/assets/3002068/e6a02466-33cb-483e-b66d-240834d278e7)
![image](https://github.com/tryAGI/LangChain/assets/3002068/64fd7499-554d-4258-9de6-5a4de9f1abf1)
it same for all models/providers but not all providers have specific settings classes like `OpenAiChatSettings`

also it will be return `usedSettings` in response:
![image](https://github.com/tryAGI/LangChain/assets/3002068/34dd22eb-fb5e-4da6-96d9-19eb543a2030)

**Comment by fernandomiletto** [comment]
@HavenDV, could you provide an example of how to implement a temperature setting of 0.4 with the GPT-3.5 Turbo model? I attempted to implement it, but it didn't work.
![Captura de tela 2024-07-03 104818](https://github.com/tryAGI/LangChain/assets/68031432/5fd2798a-8767-4f2c-bc1f-1fe118e71f62)

**Reply by danijerez** [reply]
I think the error message is clear.

> Init-only property or indexer 'OpenAiChatSettings.Temperature' can only be assigned in an object initializer, or on 'this' or 'base' in an instance constructor or an 'init' accessor.

Try doing the initialization like this:

`var settings = new OpenAiChatSettings { Temperature = 0.5 };`
